wg-naming: Add workflow for proposing recommendations (#5211)
kubernetes/community: wg-naming/workflow.md

An ADR must include:

- Groups responsible for implementing the change
- The scope of the change in the Kubernetes project, as well as downstream
  implications

ADRs must be opened with a `/hold` to give an opportunity to seek approval
with the governance groups that with be responsible for implementation.

After opening an ADR pull request, WG Naming leads
should:

- Reply to any mailing list threads about the recommendation with a link to the
  newly-opened PR, and CC any stakeholder groups
- Place the recommendation on the next meeting's agenda for review

## Approval

_This approval process is still under discussion, so here we will list out some
frequently-asked questions from our discussions thus far._

### What if a recommendation requires a KEP?

ADRs should remain on hold until scoped area agrees with the direction.

### What do we do when stakeholders disagree with a recommendation?

Again, do not merge a recommendation until code owners from the scoped area
agree to it.

### General guidance

- SIG Architecture records decisions to "...not make the mistakes we made in
  the past"
- Don’t block recording a recommendation just because everything isn’t wholly

>   figured out; let recommendations guide future work

Reviewer: we might want to clarify this point... "everything isn’t wholly figured out" is really vague :-)

I'd suggest something specific like:

> Don’t block recording a recommendation on a plan to remediate all existing uses; once the direction is agreed on by the code owners from the scoped area, a recorded recommendation has value in guiding new/future work

### Requirements

- ADR is on hold until approved by scoped areas (SIG Architecture, SIG Docs)
- WG Naming lead merges ADR
- WG Naming lead proposes ADR to Steering for official approval

> - Steering approves

Reviewer 2: I'm probably missing something here, so please feel free to point me to a link/doc if this was discussed before :)

What would be the difference between merge vs "official approval"?

Reviewer 3: I'd say skip 3 and 4. just tag steering, do a lazy consensus timeout and merge.

  implications)

ADRs must be opened with a `/hold` to give an opportunity to seek approval
with the governance groups that with be responsible for implementation.

After a recommendation ADR pull request has been opened, WG Naming Leads
should:

- Reply to any mailing list threads about the recommendation with the
  newly-opened PR, additionally including any stakeholder groups
- Place the recommendation on the next meeting's agenda for review

## Approval

_This approval process is still under discussion, so here we will list out some
frequently-asked questions from our discussions thus far._

### What to do when recommendation requires a KEP?

ADRs should remain on hold until scoped area agrees with the direction.

### What do we do when stakeholders disagree with a recommendation?

Again, do not merge a recommendation until code owners from the scoped area
agreed to it.

### General guidance

- SIG Architecture records decisions to "...not make the mistakes we made in
  the past"
- Don’t block recording a recommendation just because everything isn’t wholly
  figured out; let recommendations guide future work

### Requirements

> - ADR is on hold until approved by scoped areas (Arch, Docs)

Reviewer 4: ```suggestion
- ADR is on hold until approved by scoped areas (SIG Architecture, SIG Docs)
```

> - WG Naming Lead merges ADR

Reviewer 4: ```suggestion
- WG Naming lead merges ADR
```

> - ADR is proposed to Steering for official approval

Reviewer 4: ```suggestion
- WG Naming lead proposes ADR to Steering for official approval
```

- Steering approves

## Implementation

> - WG Naming Leads solidify accepted recommendation in some canonical location

Reviewer 4: ```suggestion
- WG Naming leads record accepted recommendation in a canonical location (TBD)
```

  (for example, a style guide)

> - Scoped area is now responsible for implementation

Reviewer 4: ```suggestion
- Areas in scope are now responsible for implementation
```
